pkg/awsutils/awsutils.go

```diff
@@ -1296,16 +1388,42 @@ func (cache *EC2InstanceMetadataCache) waitForENIAndIPsAttached(eni string, want
 }
 
 // DeallocIPAddresses allocates numIPs of IP address on an ENI
-func (cache *EC2InstanceMetadataCache) DeallocIPAddresses(eniID string, ips []string) error {
+func (cache *EC2InstanceMetadataCache) DeallocIPAddresses(eniID string, ips []string, ignoreCachedPDflag bool) error {
 	log.Infof("Trying to unassign the following IPs %s from ENI %s", ips, eniID)
-	var ipsInput []*string
-	for _, ip := range ips {
-		ipsInput = append(ipsInput, aws.String(ip))
+	ipsInput := aws.StringSlice(ips)
+
+	input := &ec2.UnassignPrivateIpAddressesInput{}
+	if !cache.useIPv4PrefixDelegation || ignoreCachedPDflag {
+		input = &ec2.UnassignPrivateIpAddressesInput{
+			NetworkInterfaceId: aws.String(eniID),
+			PrivateIpAddresses: ipsInput,
+		}
+	} else {
+		input = &ec2.UnassignPrivateIpAddressesInput{
```

Comment:
why we have both DeallocIPAddresses & DeallocPrefixAddresses and DeallocIPAddresses contains same logic as DeallocPrefixAddresses for prefix

Reply:
So if PD is enabled but we are freeing Secondary IP [this can happen during upgrades/knob disable to enable] then we set "ignoreCachedPDflag" so that we free the secondary IP.